bootloader/src/pyi_splash.c

```diff
@@ -42,6 +48,571 @@ typedef struct _ipc_message_head {
+                           ntohl(splash->wnd_height),
+                           &splash->bitmap,
+                           &bitmapinfo,
+                           DIB_RGB_COLORS);
+
+    if (lines_read != ntohl(splash->wnd_height)) {
+        FATAL_WINERROR("SetDIBits",
+                       "Could not copy all %d lines, but only %d.\n",
+                       ntohl(splash->wnd_height),
+                       lines_read);
+        return NULL;
+    }
+
+    return splash_bitmap;
+}
+
+/*
+ * Create a new bitmap on which the text is displayed.
+ * The bitmap has the size of text_rect and has per-pixel
+ * transparency. It uses the system default font.
+ *
+ * The parameter bitmap_header receives the BITMAPINFOHEADER of
+ * the created bitmap and the function returns a handle to
+ * the newly created bitmap.
+ */
+static HBITMAP
+win32HandleText(SPLASHWINDOWSTATE *pState, char *text_msg,
+                BITMAPINFOHEADER *bitmap_header)
+{
+    HDC hdc_text;
+    HBITMAP hbitmap, old_hbitmap;
+    HFONT old_text;
+    RECT text_rect = {0};
+    DWORD *pPixels = 0;
+    int pos;
+
+    /* Create DC for text bitmap */
+    hdc_text = CreateCompatibleDC(NULL);
+
+    /* Set the style of the text */
+    SetTextColor(hdc_text, ntohl(pState->splash_header->wnd_txt_clr));
+    SetBkColor(hdc_text, 0x00000000);
+    SetBkMode(hdc_text, OPAQUE);
+
+    /* Create a new bitmap for the text */
+    bitmap_header->biSize = sizeof(BITMAPINFOHEADER);
+    bitmap_header->biWidth = ntohl(pState->splash_header->wnd_txt_rect.right) -
+                             ntohl(pState->splash_header->wnd_txt_rect.left);
+    bitmap_header->biHeight = ntohl(pState->splash_header->wnd_txt_rect.bottom) -
+                              ntohl(pState->splash_header->wnd_txt_rect.top);
+    bitmap_header->biPlanes = 1;
+    bitmap_header->biBitCount = 32;
+    bitmap_header->biCompression = BI_RGB;
+
+    hbitmap = CreateDIBSection(hdc_text,
+                               (const BITMAPINFO *) bitmap_header,
+                               DIB_RGB_COLORS,
+                               (void **) &pPixels,
+                               NULL,
+                               0);
+
+    old_hbitmap = (HBITMAP) SelectObject(hdc_text, hbitmap);
+    old_text = SelectObject(hdc_text, GetStockObject(SYSTEM_FONT));
+
+    /* Create a rect for text clipping */
+    text_rect.right = bitmap_header->biWidth;
+    text_rect.bottom = bitmap_header->biHeight;
+
+    /* Draw text on the new bitmap */
+    DrawText(hdc_text,
+             pyi_win32_utils_from_utf8(NULL, text_msg, 0),
+             (int) strlen(text_msg),
+             &text_rect,
+             DT_TOP | DT_LEFT);
+
+    /*
+     * DrawText does not work well with an alpha channel,
+     * so the alpha value for all pixels with color information must be set manually.
+     */
+    for (pos = 0; pos < bitmap_header->biWidth * bitmap_header->biHeight; pos++) {
+        if (pPixels[pos] > 0x00000000) {
```

Comment:
There seems to be a bug here. One cannot draw text in the color 0x000000 (=black), since the this condition is not met